### PR TITLE
`azurerm_data_factory_linked_service_sql_server`: add `key_vault_connection_string` argument

### PR DESCRIPTION
**Pull request by favoretti**
Fixes https://github.com/terraform-providers/terraform-provider-azurerm/issues/10758#issuecomment-856827092

```
$ TF_ACC=1 go test -v ./azurerm/internal/services/datafactory -timeout=1000m -run='TestAccDataFactoryLinkedServiceSQLServer'
2021/06/08 18:59:44 [DEBUG] not using binary driver name, it's no longer needed
2021/06/08 18:59:44 [DEBUG] not using binary driver name, it's no longer needed
=== RUN   TestAccDataFactoryLinkedServiceSQLServer_basic
=== PAUSE TestAccDataFactoryLinkedServiceSQLServer_basic
=== RUN   TestAccDataFactoryLinkedServiceSQLServer_PasswordKeyVaultReference
=== PAUSE TestAccDataFactoryLinkedServiceSQLServer_PasswordKeyVaultReference
=== RUN   TestAccDataFactoryLinkedServiceSQLServer_ConnectionStringKeyVaultReference
=== PAUSE TestAccDataFactoryLinkedServiceSQLServer_ConnectionStringKeyVaultReference
=== CONT  TestAccDataFactoryLinkedServiceSQLServer_basic
=== CONT  TestAccDataFactoryLinkedServiceSQLServer_ConnectionStringKeyVaultReference
=== CONT  TestAccDataFactoryLinkedServiceSQLServer_PasswordKeyVaultReference
--- PASS: TestAccDataFactoryLinkedServiceSQLServer_basic (168.50s)
--- PASS: TestAccDataFactoryLinkedServiceSQLServer_PasswordKeyVaultReference (309.00s)
--- PASS: TestAccDataFactoryLinkedServiceSQLServer_ConnectionStringKeyVaultReference (309.43s)
PASS
ok  	github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/datafactory	310.922s
```

### CODE REVIEW

File: azurerm/internal/services/datafactory/data_factory_linked_service_sql_server_resource.go [hunk 1]
```diff
@@ -55,11 +54,36 @@ func resourceDataFactoryLinkedServiceSQLServer() *pluginsdk.Resource {
 
 			"connection_string": {
 				Type:             pluginsdk.TypeString,
-				Required:         true,
+				Optional:         true,
+				AtLeastOneOf:     []string{"connection_string", "key_vault_connection_string"},
+				ConflictsWith:    []string{"key_vault_connection_string"},
```
Comment (magodo) [inline azurerm/internal/services/datafactory/data_factory_linked_service_sql_server_resource.go:59]: Do you mean `ExactlyOneOf`?
```suggestion
				ExactlyOneOf:     []string{"connection_string", "key_vault_connection_string"},
```

File: azurerm/internal/services/datafactory/data_factory_linked_service_sql_server_resource.go [hunk 2]
```diff
@@ -55,11 +54,36 @@ func resourceDataFactoryLinkedServiceSQLServer() *pluginsdk.Resource {
...
 				DiffSuppressFunc: azureRmDataFactoryLinkedServiceConnectionStringDiff,
 				ValidateFunc:     validation.StringIsNotEmpty,
 			},
 
+			"key_vault_connection_string": {
+				Type:          pluginsdk.TypeList,
+				Optional:      true,
+				AtLeastOneOf:  []string{"connection_string", "key_vault_connection_string"},
+				ConflictsWith: []string{"connection_string"},
```
Comment (magodo) [inline azurerm/internal/services/datafactory/data_factory_linked_service_sql_server_resource.go:68]: Do you mean `ExactlyOneOf`?

File: azurerm/internal/services/datafactory/data_factory.go [hunk 3]
```diff
@@ -234,6 +234,24 @@ func expandAzureKeyVaultPassword(input []interface{}) *datafactory.AzureKeyVault
...
+			parameters["linked_service_name"] = v["referenceName"].(string)
+		}
+	}
+
+	parameters["secret_name"] = input["secretName"]
+
+	return []interface{}{parameters}
+}
+
 func flattenAzureKeyVaultPassword(secretReference *datafactory.AzureKeyVaultSecretReference) []interface{} {
```
Comment (magodo) [inline azurerm/internal/services/datafactory/data_factory.go:255]: Should we also rename method to `flattenAzureKeyVaultSecretReference`?

**Review by magodo — CHANGES_REQUESTED** [review]
@favoretti I've left some minor comments inline, over all it LGTM!

File: azurerm/internal/services/datafactory/data_factory_linked_service_sql_server_resource.go (hunk 1 above)
Comment (favoretti) [reply azurerm/internal/services/datafactory/data_factory_linked_service_sql_server_resource.go:59]: Oh yes, thank you. Got lost in all those :) will push in a sec.

**Comment by magodo** [comment]
@favoretti LGTM 👍
image

**Pull request merged**